### ISSUE TITLE
Better Chat Virtualization Config

### ISSUE THREAD

**Issue by kidroca** [Expensify/Expensify.cash#2629]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
- Chat works the same way or faster while consuming significantly less memory
- Scrolling to the past is faster

## Actual Result:
- The more chat history grows the more memory it takes
- Scrolling to the top of the list adds data message by message instead of a whole page

## Action Performed:
1. Have a chat with 200+ messages
2. Inspect the DOM element where chat actions are appended
3. Observer performance information in dev tools
4. Observe total dom nodes and memory consumption. 
5. Scroll up down and observe the information in 2 and 3 again
6. Observe how more DOM nodes are added 1-2 at a time

#### Below is a video link showing:
- [Virtual List - Before Tweaks](https://u.pcloud.link/publink/show?code=XZj011XZpAbAd1bzyCLGEWUPTP6KWBCC58mk)
    - more than 10MB and could not be hosted in github
- about 4.5k DOM nodes once a chat is fully initialized
    - 116MB of memory consumed
    - 184 Action Report Items rendered
- total nodes growing to as much as 8.5k during scrolling
   - 193MB of memory consumed
   - 327 Action Report Items rendered

#### As a comparison a chat with just one message will have
- about 425 (0.4k) DOM nodes once initialized
- 38MB of memory consumed

#### A chat with 3 (web) pages of messages will have
- about 1.3k DOM nodes once initialized
- 43MB of memory
- 27 Action Report Items rendered

## Conclusion 

Chat history would fill more and more memory and Virtualization is utilized only after rendered content is more than 12 pages of chat actions. Web pages which tend to fit more information than a mobile device page/screen.

## Workaround:
Not a bug.
Can the user still use Expensify without this being fixed? - Yes

## Platform:
Where is this issue occurring?

 - [x] Web
 - [x] iOS
 - [x] Android
 - [x] Desktop App
 - [x] Mobile Web

**Version Number:** 1.0.33
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

#### A video showing how current Virtualization works: memory, nodes, cpu usage
https://u.pcloud.link/publink/show?code=XZj011XZpAbAd1bzyCLGEWUPTP6KWBCC58mk
 - this is the same one as in the Details section

#### Secondary: A video showing scrolling to the past problems on Android

https://user-images.githubusercontent.com/12156624/116585746-5fba6c80-a921-11eb-9642-735d0d97bfa1.mp4

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by kidroca** [comment]
Here is the same test performed after some tweaks to virtualization settings for the `BaseInvertedFlatList` component

https://u.pcloud.link/publink/show?code=XZ2Y11XZVLUsJ5hVwduxs1ujmf5947ukrFnk

- about 1.8k DOM nodes once a chat is fully initialized
    - 60MB of memory consumed
    - 59 Action Report Items rendered
- total nodes growing to as much as 4.4k during scrolling
    - 81MB of memory consumed
    - 137 Action Report Items rendered

Between the two videos you can also notice 
- on the "before" video after you're done scrolling - CPU usage is kept high as items are added one by one
- on the "after" video you can see that CPU usage drops shortly after you stop scrolling

There are no noticeable speedups or slow downs. 
Scrolling fast would review blank content in both cases - this can be addressed separately

On mobile (Android) you can also notice a slight improvement and scrolling to the past working a lot better.

https://user-images.githubusercontent.com/12156624/116590986-07866900-a927-11eb-8cf7-2c3ef6a62a67.mp4

You can notice problems scrolling back to recent messages, but this happens without my changes as well 

https://user-images.githubusercontent.com/12156624/116591455-82e81a80-a927-11eb-9221-23400169d726.mp4

**Comment by kidroca** [comment]
The above shows that you can roughly cut memory usage in half and reduce CPU time as well, with no tradeoffs to quality

**Comment by marcaaron** [comment]
Thanks for documenting this. Any thoughts on how we should address it?

**Comment by kidroca** [comment]
Yes, I've basically just changed this:

```jsx
    maxToRenderPerBatch={10} // previously 1
    windowSize={5} // previously 15
```

You can easily see the effects from `maxToRenderPerBatch` when it was `1` when you scroll up and hit the virtualization limit items would be added one by one with a bit of a delay - if you observe the metrics you'll see this actually keeps the CPU busy. 
This value configures how many items should be appended at a time once you get near to the `windowSize` limit

`10` might not be the best value, but it's an improvement.

The freed memory comes mainly from`windowSize` - this is how many screens of data should be pre-rendered 
Usually in a chat you will be near the bottom of the list - on most recent messages. 
Even if you have 10-20 screens pre-rendered if you try to go back you'll see some blanks spots
Reducing this to 5 I could only see an improvement of responsiveness when I tried to scroll fast

The blank spaces that happen when you scroll fast can be covered by something as a skeleton item (placeholders) and only start to render actual items when the scrolling velocity is slow enough to accommodate this - but I don't have this in mind for the current ticket. This and the actual render performance of the flat list items are separate tasks

My proposal would be to try and tweak these settings further and apply the best value for virtualization
It might be a good idea to have separate setting for desktop/web and mobile:
- desktop/web aren't having much problems with the current settings
- mobile (particularly native) are having a hard time
Since there are already platform specific components different settings can be provided based on the `index` file

While I work on this I'll collect more data that should help with further improvements

**Comment by marcaaron** [comment]
>The blank spaces that happen when you scroll fast can be covered by something as a skeleton item (placeholders) and only start to render actual items when the scrolling velocity is slow enough to accommodate this - but I don't have this in mind for the current ticket. 

This sounds like a different issue.

>This and the actual render performance of the flat list items are separate tasks

If by "this" you mean reconfiguring the window options then I'm not sure if I agree with this e.g. if the items themselves were cheaper to render then there might be nothing wrong with the config. So the improvements seem separate, but related to me. Which one would you say is most important and should we prioritize one over the other?

**Comment by marcaaron** [comment]
FWIW, with the config options you've shared I'm seeing some pretty weird stuff on desktop and the chat messages jump around wildly while scrolling back. I'll have to test if the same thing happens on native. But the solution to make this platform dependent seems reasonable as we are likely to have less resources to work with on the native side.

**Comment by kidroca** [comment]
> If by "this" you mean reconfiguring the window options

I mean the Skeletons/Placeholders 

The current ticket is about reconfiguring the window options

**Comment by marcaaron** [comment]
Actually, the weirdness might have just been my stupid mouse scroll wheel lol

**Comment by marcaaron** [comment]
Thanks, so between optimizing the window sizes and batches and optimizing the individual items to render is there one we should focus on more than the other?

**Comment by marcaaron** [comment]
This does seem to improve the chat switching time on web as we are not queuing up as many components that are out of view so I think we can try it out?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mitchexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by kidroca** [comment]
Finding good Virtualization settings should cut memory usage in half and would be quick to address here

Optimizing `renderItem` cannot be done as single handedly and would probably come from a couple of different updates

Even `renderItem` becomes so fast it'll still consume a lot of memory for the pre-rendered stuff

**Comment by kidroca** [comment]
> Which one would you say is most important and should we prioritize one over the other?

I think addressing Chat Virtualization Config would net the most improvements, as you've noticed - less items to pre-render between chat switches is a performance gain as well as a hefty cut to memory used

**Comment by marcaaron** [comment]
>You can notice problems scrolling back to recent messages, but this happens without my changes as well

Personally, I'd rather address this issue as it looks almost worse than the reduced performance from having a large window size, but is more likely to affect large chats which not everyone has or will be scrolling through... yet! Anyways will let this one go through normal triage. 

It's tough to say what the right balance here should be. I'll let others decide how to interpret this one.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cdraeger11 (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by kidroca** [comment]
@marcaaron 

> Personally, I'd rather address this issue as it looks almost worse than the reduced performance from having a large window size

I just want to point out this already happens regardless of the changes I've proposed, you just need to scroll 15 pages to the past instead of 5, so by increasing window size we're just covering the issue

---

Did you notice how the content - the actual message, of the action is rendered with a bit of a delay

https://user-images.githubusercontent.com/12156624/117106166-0d95a300-ad88-11eb-9a8c-ed5d7bb26d34.mp4

Calculating the correct item size and scroll position depends on the content. Because the content is delayed you'd see these shifts.

Good news: While I was metering `Onyx.get` I noticed that when the caching solution was applied there were no longer blank actions - the content was rendered significantly faster. The changes we're about to make to Onyx would help with this

##### Chat content appears almost instantly with Onyx cache

https://user-images.githubusercontent.com/12156624/117108544-07a1c100-ad8c-11eb-90ec-406d150e5438.mp4

There a still some shifts happening when you scroll back to recent history, but not as much as before

---

Because item size is only available post render the content would always shift as items are added to the list - particularly when items are added to the bottom. When items are added to the top even if they expand after being added they just push content up - the current scroll position is unaffected, but when virtualization kicks in and items are added to the bottom they cause some shifts as they push content up as well
- content is pushed up as this is the natural append direction for an inverted list
- if item height can be determined before appending it - it can be put in the exact place and won't cause content to shift

As `renderItem` speed increases this shift effect would be reduced as seen by the video above, but it might be impossible to achieve such a fast speed that it no longer happens

 I think `getItemLayout` needs to be implemented for native as well - the closer the value returned is to the actual rendered height the less shifts would occur.

Perhaps another ticket can be opened and we work on that first?

**Comment by kidroca** [comment]
@marcaaron 
Something that got inspired by the `onLayout` change made here: https://github.com/Expensify/Expensify.cash/pull/2684/files#diff-3d3d60df30d1761a83ff8413b06ecef63fcd0084053cd70d5686098e71d87362R368

Since you're going to have a `onLayout` callback for each item anyway - you can capture the item index and dimensions. Then `getItemLayout` can return the captured layout - similarly to what is done in web, but without having to pre-render the item to measure it

This should provide an optimization such that render item continues to work like it currently does, as you scroll up more items are rendered, but then when you scroll down, since you're returning to content that was already rendered you know the exact size for each item and can return that through `getItemLayout` - this should prevent the content shift that happens when you scroll down

**Comment by marcaaron** [comment]
>this already happens regardless of the changes I've proposed, you just need to scroll 15 pages to the past instead of 5, so by increasing window size we're just covering the issue

That's true, but we are going to maybe highlight this issue more with this fix. I guess it doesn't really matter much. We'll need to do something about this eventually anyway. In the back of my mind is that we're also discussing the possibility of creating our own `FlatList` which will make tweaks like the ones you're proposing less attractive as a short term fixes.

>Calculating the correct item size and scroll position depends on the content. Because the content is delayed you'd see these shifts.

I did notice that! And that does make sense. Happy to hear Onyx changes will help there!

>I think getItemLayout needs to be implemented for native as well - the closer the value returned is to the actual rendered height the less shifts would occur. Perhaps another ticket can be opened and we work on that first?

That all makes sense. I suppose my reluctance is in whether to continue investing into working around `FlatList` instead of writing our own if the cards fall that way. 

Anyways, happy to let this go through triage. Thanks for all the extra info! I think there's enough context here for someone to make a decision.

**Comment by kidroca** [comment]
@marcaaron 
Sorry one last thing regarding implementing a custom `FlatList` for E.cash 
These are `VirtualList` props so if the custom flat list wraps a `VirtualList` it can use the same configuration

**Comment by marcaaron** [comment]
Thanks, that's good to know. And while it's unknown at this time, one possibility is that we will leave `VirtualizedList` behind and build something even lower level around `ScrollView`.

**Comment by kidroca** [comment]
> Thanks, that's good to know. And while it's unknown at this time, one possibility is that we will leave `VirtualizedList` behind and build something even lower level around `ScrollView`.

You'll get nothing more from `ScrollView` but just reinvent the wheel

https://reactnative.dev/docs/scrollview
> This is where FlatList comes into play. FlatList renders items lazily, when they are about to appear, and removes items that scroll way off screen to save memory and processing time.

Whatever you do you'll still need the above optimization and VList/FlatList is just a ScrollView that gives you this out of the box

If you're not satisfied with VirtualList, you won't be able to achieve much more with the ScrollView and just JS code - probably less because of all the JS code managing the complex layout you need - you'll need to go full native and create a base native component. Or find a library...

**Comment by marcaaron** [comment]
Thanks, let's maybe have this conversation elsewhere.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @strepanier03 (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by strepanier03** [comment]
I'm not going to be able to get to this before leaving for the day and I'm out of office for the next 6 business days. I'm reassigning and appreciate the help of the next triager, thank you 🙌

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by stephanieelliott** [comment]
Moving this to `Monthly` priority while we discuss this internally!

**Comment by marcaaron** [comment]
Going to investigate these changes to see if I can prove they will modestly improve chat switching / cold boot times.

**Comment by marcaaron** [comment]
Here is the following test I ran on an iOS device...

1. Create a new account 
2. Add two chats each with 500 comments
3. Switch from chat A to chat B

**Observation before changes applied:**

When switching from chat A to chat B there's a loading spinner and finally only the first 10 messages load followed by the others eventually. The data for each chat is stored locally despite this happening.

**When updated to use the recommended config:**

**iOS:** There is no significant difference observed and the same very bad performance issues exists with or without the changes.

**Web:** Performance seems not worse per se but also bad in different way as the slow rendering issue causes each "chunk" of messages to appear like so instead of one at a time (both are not ideal IMO)...

![2021-07-13_12-20-50 (1)](https://user-images.githubusercontent.com/32969087/125532890-5ff147af-fc30-4d95-9400-b04dc963903a.gif)

The original recommendations seem maybe like they had some benefit on Android, but still, seems to not improve chat switching times much. Ideally the config would be set like this:

1. Figure out **why** this takes so long to render and improve rendering speed
2. Load the reasonable number of chats to fill the visible area `maxToRenderPerBatch` so there is never any "empty space" that appears to fill up when we first open a chat. See this example
![2021-07-13_12-30-33 (1)](https://user-images.githubusercontent.com/32969087/125533488-44ba1c25-f00a-4188-b0b5-ade315152b60.gif) which is what we have now.
3. Modify the `windowSize` to prevent blank spaces while scrolling

I'm going to close this issue and break these findings out into some new issues. We can revisit any tweaks to the config as a solution to some specific problem e.g. blank spaces. Our current focus is on improving chat switching + message rendering speeds.
